File: C++/Linked List/Doubly Linked List/DoublyLL04.cpp
/*
===============================================================================
Program Name    : Doubly Linear Linked List - Insert First
Program Number  : 004
File Name       : DoublyLL04.cpp

Description     : This program demonstrates insertion of a node
                  at the beginning of a Doubly Linear Linked List
                  using C++.

                  The InsertFirst() function creates a new node
                  and inserts it as the first node of the linked
                  list while maintaining proper next and previous
                  links.

Date            : July 2026
Language        : C++
Author          : Atharv Tushar Bhosale
===============================================================================

Concepts Used:
1. Class
2. Object-Oriented Programming (OOP)
3. Self-Referential Class
4. Dynamic Memory Allocation using new
5. Pointer Manipulation
6. Doubly Linked List
7. Constructor

Key Learning Points:
- Create nodes dynamically.
- Insert node at the beginning.
- Update next and previous links.
- Maintain head pointer.
- Display linked list.
- Count total nodes.

===============================================================================
*/

#include<iostream>
using namespace std;

class node
{
public:
    int data;
    node *next;
    node *prev;
};

typedef class node NODE;
typedef class node *PNODE;

class DoublyLL
{
private:
    PNODE first;
    int iCount;

public:
    DoublyLL();

    void Display();
    int Count();

    void InsertFirst(int);
};

DoublyLL::DoublyLL()
{
    first = NULL;
    iCount = 0;
}

void DoublyLL::Display()
{
    PNODE temp = first;

    while(temp != NULL)
    {
        cout<<"NULL <=> | "<<temp->data<<" | <=> ";
        temp = temp->next;
    }

    cout<<"NULL\n";
}

int DoublyLL::Count()
{
    return iCount;
}

void DoublyLL::InsertFirst(int no)
{
    PNODE newn = new NODE;

    newn->data = no;
    newn->next = NULL;
    newn->prev = NULL;

    if(first == NULL)
    {
        first = newn;
    }
    else
    {
        newn->next = first;
        first->prev = newn;
        first = newn;
    }

    iCount++;
}

int main()
{
    DoublyLL dobj;

    dobj.InsertFirst(51);
    dobj.InsertFirst(21);
    dobj.InsertFirst(11);

    dobj.Display();

    cout<<"Number of nodes are : "
        <<dobj.Count()<<endl;

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

NULL <=> | 11 | <=> NULL <=> | 21 | <=> NULL <=> | 51 | <=> NULL

Number of nodes are : 3

===============================================================================
Memory Representation:
===============================================================================

first
  |
  v

NULL <- [11] <=> [21] <=> [51] -> NULL

===============================================================================
Algorithm:
===============================================================================

1. Create a new node.
2. Store data in the node.
3. If linked list is empty, make new node as first.
4. Otherwise:
   - Connect new node to current first node.
   - Update previous link of first node.
   - Move first pointer to new node.
5. Increment node count.
6. Display the linked list.

===============================================================================
Time Complexity:
===============================================================================

InsertFirst() : O(1)
Display()     : O(n)
Count()       : O(1)

Space Complexity : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 004 demonstrates insertion at the beginning of a Doubly
Linear Linked List. The InsertFirst() function efficiently inserts
a node in constant time by updating the first pointer and maintaining
proper next and previous links.

===============================================================================
*/

File: Java/Linked List/Doubly Linked List/DoublyLL04.java
/*
===============================================================================
Program Name    : Doubly Linear Linked List - Display
Program Number  : 004
File Name       : DoublyLL04.java

Description     : This program demonstrates the implementation of the
                  Display() operation in a Doubly Linear Linked List
                  using Java. The Display() function traverses the
                  linked list from the first node to the last node
                  and displays each element.

Date            : July 2026
Language        : Java
Author          : Atharv Tushar Bhosale
===============================================================================

Concepts Used:
1. Class and Object
2. Self-Referential Class
3. Doubly Linear Linked List
4. Traversing Linked List
5. while Loop
6. Pointer Manipulation
7. Encapsulation

Key Learning Points:
- Traverse the linked list.
- Display each node's data.
- Stop traversal when NULL is reached.
- Understand sequential node access.

===============================================================================
*/

class node
{
    public int data;
    public node next;
    public node prev;

    node(int no)
    {
        this.data = no;
        this.next = null;
        this.prev = null;
    }
}

class DoublyLL
{
    private node first;
    private int iCount;

    public DoublyLL()
    {
        first = null;
        iCount = 0;
    }

    public void Display()
    {
        node temp = first;

        while(temp != null)
        {
            System.out.print("| " + temp.data + " | <=> ");
            temp = temp.next;
        }

        System.out.println("NULL");
    }
  
    public void InsertLast(int iNo)
    {
        node newn = new node(iNo);
        node temp = first;

        if(first == null)
        {
            first = newn;
        }
        else
        {
            while(temp.next != null)
            {
                temp = temp.next;
            }

            temp.next = newn;
            newn.prev = temp;
        }

        iCount++;
    }
}

class DoublyLL04
{
    public static void main(String A[])
    {
        DoublyLL dobj = new DoublyLL();

        dobj.InsertLast(11);
        dobj.InsertLast(21);
        dobj.InsertLast(51);
        dobj.InsertLast(101);

        dobj.Display();
    }
}

/*
===============================================================================
Expected Output:
===============================================================================

| 11 | <=> | 21 | <=> | 51 | <=> | 101 | <=> NULL

===============================================================================
Memory Representation:
===============================================================================

first
 │
 ▼

NULL <- [11] <=> [21] <=> [51] <=> [101] -> NULL

===============================================================================
Algorithm:
===============================================================================

1. Create a temporary pointer.
2. Point it to the first node.
3. Traverse while the pointer is not NULL.
4. Display each node's data.
5. Move to the next node.
6. Stop when NULL is reached.

===============================================================================
Time Complexity:
===============================================================================

Display() : O(n)

Space Complexity : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 004 demonstrates the Display() operation in a Doubly Linear
Linked List using Java. It traverses the linked list from beginning
to end and displays all node values in sequence.

===============================================================================
*/
